### PR TITLE
internal interface: pf.flows._save

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+@pytest.mark.usefixtures(
+    "use_secrets_config_file", "recording_injection", "setup_local_connection", "install_custom_tool_pkg"
+)
+@pytest.mark.sdk_test
+@pytest.mark.e2etest
+class TestFlowSave:
+    def test_pf_save_simple(self):
+        pf = PFClient()
+        # or follow spec: import promptflow as pf? seems that we don't have such interface now.
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py:37]: not this: import promptflow as pf

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            # this should be entry instead of flow as we don't have a flow now in concept level before saving
+            entry="hello:hello_world",
+            # question: should the requirements be related to code?
+            # or should we copy it based on current working directory?
+            requirements=f"{TEST_ROOT}/test_configs/functions/requirements",
+            image="python:3.8-slim",
+            # all content in code will be copied
+            code=f"{TEST_ROOT}/test_configs/functions/hello_world",
+            # what if the signature doesn't match the code? should we do any validation?
+            signature={
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py:50]: need to do basic validation like the input name should match

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        pf = PFClient()
+        # or follow spec: import promptflow as pf? seems that we don't have such interface now.
+        pf.save(
+            # should we support save to a yaml file and do not copy code?
+            path=f"{FLOWS_DIR}/saved/hello_world",
+            # this should be entry instead of flow as we don't have a flow now in concept level before saving
+            entry="hello:hello_world",
+            # question: should the requirements be related to code?
+            # or should we copy it based on current working directory?
+            requirements=f"{TEST_ROOT}/test_configs/functions/requirements",
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py:45]: python_requirements?

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                    }
+                },
+                "outputs": {
+                    "answer": {
+                        "type": "str",
+                        "description": "The answer",
+                    }
+                },
+            },
+            # this field looks a little unnecessary. If we want it for pf flow test, seems that we should use it to
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py:64]: used for swagger gen

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            # set default values for flow inputs; if we want it for run create, we should create a jsonl.
+            input_sample={"text": "promptflow"},
+            # or we should infer it from entry?
+            language="python",
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py:68]: optional

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                    }
+                },
+                # what about use this instead of a separate init_signature?
+                "init": {
+                    "the_model": {
+                        "type": "str",
+                    }
+                },
+            },
+            force=True,
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_save.py:104]: not expose this in first version

File: src/promptflow-devkit/promptflow/_sdk/operations/_flow_operations.py [hunk 7]
```diff
@@ -906,3 +912,123 @@ def _generate_flow_meta(
...
+                    raise UserErrorException(
+                        f"Provided signature of {key} for entry {entry} does not match the actual signature.\n"
+                        f"Provided: {', '.join(original_signature[key].keys())}\n"
+                        f"Actual: {', '.join(entry_meta[key].keys())}\n"
+                    )
+
+        return original_signature
+
+    @monitor_operation(activity_name="pf.flows._save", activity_type=ActivityType.INTERNALCALL)
+    def _save(
```
Comment (elliotzh) [inline src/promptflow-devkit/promptflow/_sdk/operations/_flow_operations.py:955]: @D-W- use this for internal call?

File: src/promptflow-devkit/promptflow/_proxy/_python_inspector_proxy.py [hunk 8]
```diff
@@ -21,6 +25,25 @@ def get_used_connection_names(
         return executable.get_connection_names(environment_variables_overrides=environment_variables_overrides)
 
     @classmethod
-    def is_flex_flow_entry(self, entry: str) -> bool:
+    def is_flex_flow_entry(cls, entry: str) -> bool:
         """Check if the flow is a flex flow entry."""
         return isinstance(entry, str) and re.match(FlowEntryRegex.Python, entry)
+
+    @classmethod
+    def get_entry_meta(
```
Comment (elliotzh) [inline src/promptflow-devkit/promptflow/_proxy/_python_inspector_proxy.py:33]: @D-W- @YingChen1996 use this for direct meta generation? besides, if I understand correctly, we won't need to dump the json anymore?

File: src/promptflow-devkit/promptflow/_sdk/operations/_flow_operations.py [hunk 9]
```diff
@@ -875,3 +881,123 @@ def _generate_flow_meta(
...
+
+        target_flow_file = target_flow_directory / DAG_FILE_NAME
+        target_flow_directory.parent.mkdir(parents=True, exist_ok=True)
+
+        # TODO: handle ignore
+        shutil.copytree(code, target_flow_directory)
+        if python_requirements:
+            shutil.copy(python_requirements, target_flow_directory / Path(python_requirements).name)
+        if input_sample:
+            with open(target_flow_directory / "sample.json", "w", encoding=DEFAULT_ENCODING) as f:
```
Comment (wangchao1230) [inline src/promptflow-devkit/promptflow/_sdk/operations/_flow_operations.py:1000]: give sample.json a constant

File: src/promptflow-devkit/promptflow/_sdk/schemas/_flow.py [hunk 10]
```diff
@@ -65,6 +65,9 @@ class EagerFlowSchema(BaseFlowSchema):
 
     # entry point for eager flow
     entry = fields.Str(required=True)
+    inputs = fields.Dict(keys=fields.Str(), values=NestedField(FlowInputSchema), required=False)
+    outputs = fields.Dict(keys=fields.Str(), values=NestedField(FlowOutputSchema), required=False)
+    inits = fields.Dict(keys=fields.Str(), values=NestedField(FlowInputSchema), required=False)
```
Comment (wangchao1230) [inline src/promptflow-devkit/promptflow/_sdk/schemas/_flow.py:70]: inits -> init?

File: src/promptflow-core/promptflow/_core/entry_meta_generator.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+"""
+This file can generate metadata for the flow entry.
+"""
+import json
+import multiprocessing
+from pathlib import Path
+from typing import Dict, Union
+
+from promptflow._constants import DEFAULT_ENCODING, FLOW_META_JSON, FLOW_META_JSON_GEN_TIMEOUT, PROMPT_FLOW_DIR_NAME
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/_core/entry_meta_generator.py:13]: I think we need to update devkit dependency to core>=1.1.0? @brynn-code

**Review by thy09 — APPROVED** [review]

**Review by D-W- — APPROVED** [review]

**Pull request merged**